packages/glimmer-runtime/lib/compiled/opcodes/vm.ts

```diff
@@ -306,6 +313,77 @@ export class EvaluateOpcode extends Opcode {
   }
 }
 
+export class EvaluatePartialOpcode extends Opcode {
+  public type = "evaluate-partial";
+  public compiler: CompileInto & SymbolLookup;
+  public name: CompiledExpression<any>;
+
+  constructor({ name, compiler }: { compiler: CompileInto & SymbolLookup, name: CompiledExpression<any> }) {
+    super();
+    this.name = name;
+    this.compiler = compiler;
+  }
+
+  evaluate(vm: VM) {
+    let reference: PathReference<any> = this.name.evaluate(vm);
+    let name: InternedString = reference.value();
+
+    let { serializedTemplate } = vm.env.lookupPartial([name]);
+    let scanner = new Scanner(serializedTemplate, vm.env);
+    let block = scanner.scanInlineBlock(this.compiler.getSymbolTable());
+    vm.invokeBlock(block, vm.frame.getArgs());
+
+    if (!isConst(reference)) {
+      vm.pushDynamicScope();
+
+      vm.updateWith(new EvaluateUpdatingPartialOpcode({
+        compiler: this.compiler,
+        nameReference: reference,
+        name
+      }));
+    }
+  }
+
+  toJSON(): OpcodeJSON {
+    return {
+      guid: this._guid,
+      type: this.type,
+      args: [this.name.toJSON()]
+    };
+  }
+}
+
+
+export class EvaluateUpdatingPartialOpcode extends UpdatingOpcode {
```

Comment:
We don't need a specialized opcode here. We already have [`Assert`](https://github.com/tildeio/glimmer/blob/master/packages/glimmer-runtime/lib/compiled/opcodes/vm.ts#L392-L429) that does exactly this. You just give it a [`ReferenceCache`](https://github.com/tildeio/glimmer/blob/master/packages/glimmer-reference/lib/validators.ts#L275-L324) and it will take care of the rest.